File: Chapter 05/Python/section_5_6_crypto_tracker/after/core/domain/portfolio_manager.py
from ...core.ports.price_provider_port import PriceProviderPort

class PortfolioManager:
    """THE INSIDE (The Core).
    
    Contains the pure business logic for financial calculations.
    This class is fully isolated from HTTP clients and JSON parsing, 
    relying entirely on an abstract PriceProviderPort.
    """

    def __init__(self, price_provider: PriceProviderPort):
        """Initializes the manager with a price provider.
        
        Args:
            price_provider (PriceProviderPort): An adapter that satisfies 
                the get_bitcoin_price contract (Constructor Injection).
        """
        self.price_provider = price_provider

    def calculate_total_value(self, btc_amount: float) -> float:
        """Calculates the total market value of a Bitcoin holding.
        
        Args:
            btc_amount (float): The quantity of Bitcoin to evaluate.
            
        Returns:
            float: The total value in USD.
        """
        # We call the port. We don't care WHERE the price comes from.
        current_price = self.price_provider.get_bitcoin_price()
        total_value = btc_amount * current_price
        
        # Professional trace logging for developers to see the math breakdown
        print(f"[Core] Calculating: {btc_amount} BTC x ${current_price:,.2f}/BTC")
        
        return total_value

File: Chapter 05/JS/section_5_6_crypto_tracker/after/core/domain/portfolioManager.js
/**
 * CORE – Pure business logic.
 * No fetch calls, no JSON parsing. This class is fully isolated.
 */
class PortfolioManager {
    /**
     * Dependency Injection via Constructor.
     * @param priceProvider - An adapter that satisfies the PriceProvider contract.
     */
    constructor(priceProvider) {
        this.priceProvider = priceProvider;
    }

    /**
     * Calculates total value by awaiting the external price data.
     * @param btcAmount - The quantity of Bitcoin to calculate.
     */
    async calculateTotalValue(btcAmount) {
        // We must await the result from the provider
        const currentPrice = await this.priceProvider.getBitcoinPrice();
        const totalValue = btcAmount * currentPrice;

        const usd = new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD' });

        // Now currentPrice will be a number, not a Promise object
        console.log(`[Core] Calculating: ${btcAmount} BTC x ${usd.format(currentPrice)}/BTC`);
        
        return totalValue;
    }
}

module.exports = PortfolioManager;